controllers/storagecluster/consumer.go

```diff
@@ -0,0 +1,168 @@
+
+	ocsv1 "github.com/red-hat-storage/ocs-operator/api/v1"
+	providerClient "github.com/red-hat-storage/ocs-operator/services/provider/client"
+	"github.com/red-hat-storage/ocs-operator/services/provider/common"
+)
+
+var (
+	// externalOCSResources will hold the ExternalResources for storageclusters
+	// ExternalResources can be accessible using the UID of an storagecluster
+	// getExternalConfigFromProvider func sets value for this once and use same again
+	externalOCSResources = map[types.UID][]ExternalResource{}
+)
+
+type ocsConsumer struct{}
+
+// ensureCreated performs onboarding and update capacity
+func (o *ocsConsumer) ensureCreated(r *StorageClusterReconciler, instance *ocsv1.StorageCluster) error {
+
+	if !isExternalOCSProvider(instance) {
+		return nil
+	}
+
+	if instance.Status.ExternalStorage.ConsumerID == "" {
+		return onBoardConsumer(r, instance)
+	}
+
+	if instance.Spec.ExternalStorage.RequestedCapacity != nil &&
+		!instance.Status.ExternalStorage.GrantedCapacity.Equal(*instance.Spec.ExternalStorage.RequestedCapacity) {
+		return updateConsumerCapacity(r, instance)
+	}
+
+	return nil
+}
+
+// ensureDeleted performs offboarding
+func (o *ocsConsumer) ensureDeleted(r *StorageClusterReconciler, instance *ocsv1.StorageCluster) error {
+
+	if !isExternalOCSProvider(instance) {
+		return nil
+	}
+
+	return offBoardConsumer(r, instance)
+}
+
+// onBoardConsumer makes an API call to the provider cluster for onboarding
+func onBoardConsumer(r *StorageClusterReconciler, instance *ocsv1.StorageCluster) error {
+
+	// TODO: think about naming convention
+	response, err := r.providerClient.OnboardConsumer("my-ocs", getExternalStorageRequestedCapacity(instance))
+	err = r.handleOnboardConsumerErrors(instance, err)
+	if err != nil {
+		return err
+	}
+
+	if response.StorageConsumerUUID == "" || response.GrantedCapacity == "" {
+		err = fmt.Errorf("External-OCS:OnboardConsumer:response is empty")
+		r.Log.Error(err, "empty response")
+		return err
+	}
+
+	instance.Status.ExternalStorage.ConsumerID = response.StorageConsumerUUID
+	instance.Status.ExternalStorage.GrantedCapacity = resource.MustParse(response.GrantedCapacity)
+
+	return nil
+}
+
+// offBoardConsumer makes an API call to the provider cluster for offboarding
+func offBoardConsumer(r *StorageClusterReconciler, instance *ocsv1.StorageCluster) error {
+
+	_, err := r.providerClient.OffboardConsumer(instance.Status.ExternalStorage.ConsumerID)
+	return r.handleOffboardConsumerErrors(instance, err)
+}
+
+// upscaleOrDownscaleConsumer makes an API call to the provider cluster to update the capacity
+func updateConsumerCapacity(r *StorageClusterReconciler, instance *ocsv1.StorageCluster) error {
+
+	response, err := r.providerClient.UpdateCapacity(
+		instance.Status.ExternalStorage.ConsumerID,
+		getExternalStorageRequestedCapacity(instance))
+	err = r.handleUpdateConsumerCapacityErrors(instance, err)
+	if err != nil {
+		return err
+	}
+
+	if response.GrantedCapacity == "" {
```

Comment:
How can this happen if updateCapacity was successful?

Reply:
It should not, I am just making sure that the server has set the value.